### DISCUSSION TITLE
Vertx Gateway Memory Consumption Issue.

### DISCUSSION THREAD

**Discussion by shekarnode** [apiman/apiman#1342] [Help]
Hi Marc,

Trust you are doing well.

We are using vertx as api gateway for our application as I had already mentioned in one of the thread but we are facing very unusual scenario where even in the idle mode memory keeps on increasing. Right now we are running vertx gateway inside docker container with -Xms 1024m but when the system is in idle mode memory keeps on increasing and does come back to normal even if there are no requests. 

We have added one  policy to attach Transaction ID and we are storing each transaction response code in redis from that policy. 

Attaching Graphana chart of the docker container :
![graphana_01](https://user-images.githubusercontent.com/3267543/123044007-bec6ca00-d416-11eb-8e7e-854b5a94f602.png)

Docker file content :
```
FROM adoptopenjdk/openjdk8-openj9:alpine

RUN mkdir -p /app/_cache
COPY apiman-distro-vertx-1.5.7.Final /app
#COPY build/libs/openj9-howto-all.jar /app/app.jar
VOLUME /app/_cache
#EXPOSE 8080
CMD ["java","-Xmx1024m", "-Xvirtualized", "-Xshareclasses", "-Xshareclasses:name=sum", "-Xshareclasses:cacheDir=/app/_cache", "-jar", "/app/apiman-gateway.jar","--conf", "/app/configs/conf-es.json"]

```
Any suggestions on optimization of vertx gateway ?

**Comment by volkflo** [comment]
Hi @shekarnode, that looks interesting, thanks for sharing that with us.
We will have a look into it.

**Comment by msavy** [comment]
@shekarnode can you please use something like Java Mission Control (JMC) to assess the situation? It's not uncommon for modern Java garbage collectors and JVMs to fairly rapidly expand up to their XMX and then not shrink down (for performance reasons). Hence, to differentiate between that and anything problematic you will need to dig into it a little bit and see what is actually using the memory. 

JMC has a suite of tools that will allow you to peer into the heap and see what's using the memory in the application (or whether it's just being held by the JVM).

Other possible causes could be a Java memory leak or a cache that is expanding indefinitely, etc.

Let us know what you find out

**Reply by shekarnode** [reply]
Ok I will check on this and get back.

**Comment by msavy** [comment]
BTW, we don't do any kind of validation/testing on openj9. Why do you use that one?

**Reply by shekarnode** [reply]
We are trying to make gateway consume less memory

So we thought of using using this
[Vertx Doc](https://how-to.vertx.io/openj9-howto/)

What do you suggest? Should we go for adoptopenjdk or openjdk hotspot

**Reply by msavy** [reply]
You can use whatever you like, of course! But I personally have zero experience with j9, so you're on your own in terms of profiling, etc.

Without knowing a lot of information about your deployment, configuration, use-case, traffic levels, etc, I can't give you any specific advice on how to optimise your setup.

**Reply by shekarnode** [reply]
So I am using this gateway for alternative to my zuul api gateway and currently we have around 150 odd microservices running in the background which requests api gateway very frequently.  This vertx gateway calls controller which is responsible for calling back end microservices using eureka discovery. So vertx gateway is being called for each and every request. Also there is scheduler in placed using quartz which is calling gateway after every 5 min to perform certain operations. ( 20-30 operations within few minutes ). 

So this is my current setup. So I am trying to optimize this vertx gateway to use very less memory,  but not able to do it as of now. 

Could you please comment or suggest something here ?
